drlhp/pref_interface.py

```diff
@@ -151,20 +135,9 @@ def sample_seg_pair(self):
         Sample a random pair of segments which hasn't yet been tested.
         """
         segment_idxs = list(range(len(self.segments)))
-        shuffle(segment_idxs)
-        possible_pairs = combinations(segment_idxs, 2)
-        self.remaining_possible_pairs = len(list(deepcopy(possible_pairs))) - len(self.tested_pairs)
-        # print(f"Num segments: {len(self.segments)}")
-        # print(f"Remaining pairs: {self.remaining_possible_pairs}")
-        # print(f"Tested pairs: {len(self.tested_pairs)}")
-        for i1, i2 in possible_pairs:
-            i1, i2 = min(i1, i2), max(i1, i2)
-            # these should now always be in a canonical order
-            s1, s2 = self.segments[i1], self.segments[i2]
-            if (s1.hash, s2.hash) not in self.tested_pairs:
-                self.tested_pairs.add((s1.hash, s2.hash))
-                return s1, s2
-        raise IndexError("No segment pairs yet untested")
+        possible_pairs = list(combinations(segment_idxs, 2))
+        ind1, ind2 = possible_pairs[0]
```

Comment:
```suggestion
        shuffle(segment_idxs)
        ind1, ind2 = segment_idxs[:2]
```
This avoids the quadratic-time construction of all of the combinations (though you could do `next(combinations(...))` to avoid that) but also does randomization correctly (the existing code doesn't randomize at all).

You could make it even more efficient by avoiding the `shuffle` and choosing two random numbers constrained to be different as your indices, but that's probably unnecessary.

Reply:
Good call, this makes a lot more sense given the redesign of this section